sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -241,13 +258,67 @@ rcl_interfaces::msg::SetParametersResult RingOutlierFilterComponent::paramCallba
   if (get_param(p, "num_points_threshold", num_points_threshold_)) {
     RCLCPP_DEBUG(get_logger(), "Setting new num_points_threshold to: %d.", num_points_threshold_);
   }
+  if (get_param(p, "publish_excluded_points", publish_excluded_points_)) {
+    RCLCPP_DEBUG(
+      get_logger(), "Setting new publish_excluded_points to: %d.", publish_excluded_points_);
+  }
 
   rcl_interfaces::msg::SetParametersResult result;
   result.successful = true;
   result.reason = "success";
 
   return result;
 }
+
+sensor_msgs::msg::PointCloud2 RingOutlierFilterComponent::extractExcludedPoints(
+  const sensor_msgs::msg::PointCloud2 & input, const sensor_msgs::msg::PointCloud2 & output,
+  float epsilon)
+{
+  // Convert ROS PointCloud2 message to PCL point cloud for easier manipulation
+  pcl::PointCloud<pcl::PointXYZ>::Ptr input_cloud(new pcl::PointCloud<pcl::PointXYZ>);
+  pcl::PointCloud<pcl::PointXYZ>::Ptr output_cloud(new pcl::PointCloud<pcl::PointXYZ>);
+  pcl::fromROSMsg(input, *input_cloud);
+  pcl::fromROSMsg(output, *output_cloud);
+
+  pcl::PointCloud<pcl::PointXYZ>::Ptr excluded_points(new pcl::PointCloud<pcl::PointXYZ>);
+
+  for (const auto & input_point : *input_cloud) {
+    bool is_excluded = true;
+
+    for (const auto & output_point : *output_cloud) {
+      float distance = autoware::common::geometry::distance_3d(input_point, output_point);
+
+      // If the distance is less than the threshold (epsilon), the point is not excluded
+      if (distance < epsilon) {
+        is_excluded = false;
+        break;
+      }
+    }
+
+    // If the point is still marked as excluded after all comparisons, add it to the excluded
+    // pointcloud
+    if (is_excluded) {
+      excluded_points->push_back(input_point);
+    }
+  }
```

Comment:
As our internal discussion, the current implementation is quite slow. The faster way to get excluded points is extracting during [this](https://github.com/badai-nguyen/autoware.universe/blob/5640819dbe06bfc4a676e1ce4e80a81177b0344f/sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp#L101-L185)
To prevent the change from affecting current performance, temporarily it agree to add change only for debug mode.
BTW, using `kdtree.search.filter` instead could improve processing time, I guess. 
```suggestion
    pcl::search::Search<pcl::PointXYZ>::Ptr tree;
  if (output_cloud->isOrganized()) {
    tree.reset(new pcl::search::OrganizedNeighbor<pcl::PointXYZ>());
  } else {
    tree.reset(new pcl::search::KdTree<pcl::PointXYZ>(false));
  }
  tree->setInputCloud(output_cloud);
  std::vector<int> nn_indices(1);
  std::vector<float> nn_dists(1);
  for(const auto & point : input_cloud->points)
  {
    if(!tree->nearestKSearch(point, 1, nn_indices, nn_dists))
    {
      continue;
    }
    if(nn_dists[0] > epsilon)
    {
    excluded_points->points.push_back(point);
    }
  }
```

Reply:
@badai-nguyen 
Thank you so much for your suggestion and I confirmed successfully after applying the change in https://github.com/autowarefoundation/autoware.universe/pull/6483/commits/890889f48a331d165ad6309fa7067df68c4deb64
If no other concern can you approve this PR?
I will ask maintainer to aprove this PR.